sdk/storage/Azure.Storage.Internal.Avro/src/AvroParser.cs

```diff
@@ -0,0 +1,388 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Text;
+using System.Text.Json;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Core.Pipeline;
+
+#pragma warning disable SA1402 // File may only contain a single type
+
+namespace Azure.Storage.Internal.Avro
+{
+    internal static class AvroParser
+    {
+        public static async Task<byte[]> ReadFixedBytesAsync(
+            Stream stream,
+            int length,
+            bool async,
+            CancellationToken cancellationToken)
+        {
+            byte[] data = new byte[length];
+            int start = 0;
+            while (length > 0)
+            {
+                int n = async ?
+                    await stream.ReadAsync(data, start, length, cancellationToken).ConfigureAwait(false) :
+                    stream.Read(data, start, length);
+                start += n;
+                length -= n;
+
+                // We hit the end of the stream
+                if (n <= 0)
+                    return data;
+            }
+            return data;
+        }
+
+        private static async Task<byte> ReadByteAsync(
+            Stream stream,
+            bool async,
+            CancellationToken cancellationToken)
+        {
+            byte[] bytes = await ReadFixedBytesAsync(stream, 1, async, cancellationToken).ConfigureAwait(false);
+            return bytes[0];
+        }
+
+        // Stolen because the linked references in the Avro spec were subpar...
+        private static async Task<long> ReadZigZagLongAsync(
+            Stream stream,
+            bool async,
+            CancellationToken cancellationToken)
+        {
+            byte b = await ReadByteAsync(stream, async, cancellationToken).ConfigureAwait(false);
+            ulong next = b & 0x7FUL;
+            int shift = 7;
+            while ((b & 0x80) != 0)
+            {
+                b = await ReadByteAsync(stream, async, cancellationToken).ConfigureAwait(false);
+                next |= (b & 0x7FUL) << shift;
+                shift += 7;
+            }
+            long value = (long)next;
+            return (-(value & 0x01L)) ^ ((value >> 1) & 0x7fffffffffffffffL);
+        }
+
+        public static Task<object> ReadNullAsync() => Task.FromResult<object>(null);
+
+        public static async Task<bool> ReadBoolAsync(
+            Stream stream,
+            bool async,
+            CancellationToken cancellationToken)
+        {
+            byte b = await ReadByteAsync(stream, async, cancellationToken).ConfigureAwait(false);
+            return b != 0;
```

Comment:
Can we add explicit checking here - if 0 return false, if 1 return true, otherwise throw. I find it weird that the original Apache code didnt do this check

Reply:
fixed.